### ISSUE TITLE
Help with interpolating element height without stair-stepping / Floating-point layout coordinates

### ISSUE THREAD

**Issue by kraziemaniac** [ocornut/imgui#8119]
### Version/Branch of Dear ImGui:

Version 1.91.0, Branch: master

### Back-ends:

imgui_impl_win32.cpp + imgui_impl_dx12.cpp

### Compiler, OS:

Windows 11, MSVC 2019

### Details:

Hello!

[ Preface ]
I wanted to take advantage of the drag and drop feature within Imgui to essentially drag the header of a tree node onto a space between tree nodes at the same depth where a child ( ImGui::BeginChild() ) can be used as a drag drop target. I was thinking of doing this so it can be possible to drag around tree nodes to a more preferred order for the user.

I have this working properly, and things can be ordered as desired, but the issue that I'm having is that the process itself is a bit jarring for the user for a few cosmetic reasons, or at least that's to say that I'm sure this can visually be made to be less jarring.

The tree nodes are using default ItemSpacing ( 4 ), and I was hoping that I could maybe grab the ItemSpacing value and store it onto a variable where I can interpolate it, and push that as a style var where it can go down to 0 ( or near 0 ) followed by adding a child within each tree node at that level where it can also interpolate up to a desired size for the sake of UX.

[ Problem ]
The problem that I'm having though is that it looks like ImGui::ItemSize() is ultimately truncating the y values that result from ItemSpacing, which is causing a stair-stepped interpolation when changing the value over time. I know the comment for it is to align things on pixel boundaries, which I'm sure is helpful for really long lists and floating point errors, but I'm wondering if there could be some way in the future to allow for floating point values especially when it's a relatively small tree ( < 10 elements ) ?

TLDR:
Might there in the future be a way to push ImGuiStyleVar_ItemSpacing with a floating point value so it can be interpolated over time using smaller steps than whole number steps?

**Comment by kraziemaniac** [comment]
Update:

I was thinking about this, and came up with the idea that maybe I can push ImGuiStyleVar_ItemSpacing immediately to 0 in the y, and maybe I can just dynamically change the value of the child in question instead starting from the initial ItemSpacing amount, but it seems like the stair-stepping still exists in this case. I wasn't too sure what was specifically causing that sort of truncation when I was debugging through it.

I suppose at this point the question has more to do with how to size an element ( ImGui::BeginChild() in my case ) dynamically where it doesn't stair step as it's being fed an interpolated value

**Comment by veteri** [comment]
There is some min height defined somewhere even if you pass 1px, i ran into the same problem at some point trying to slide in windows.

**Comment by ocornut** [comment]
The pixel alignment are currently required for many reasons, not only related to floating point accuracy. Rendering depends on it, and some calculation (e.g. clipper) may even be affected by it. 

Ultimately I would like to support smoother zooming and experiment with alternative way to support DPI scaling, so we may alter those constraints in the future.

Would you mind posting videos of your setup so have a little more contexts to suggest out of the box alternatives?

**Comment by kraziemaniac** [comment]
![dragdropexpansion1](https://github.com/user-attachments/assets/5ac2ef61-28ab-4a84-8794-2560b4246f60)
![dragdropexpansion2](https://github.com/user-attachments/assets/99dd1c8a-6e84-426c-b86e-439fad7f2acb)
![dragdropexpansion3](https://github.com/user-attachments/assets/9221d39c-255e-49c4-aec8-dd086f79d8a1)

I'm not able to post a video so easily at the moment, but here are a few screenshots of the relevant parts. First there's the tree nodes right next to each other, then in the next picture I'm trying to drag the top tree node to an in-between drop target but without any adjustment to the spacing which I think illustrates that the spacing should be expanded to make this a smoother experience. In the last photo, the spacing is expanded a bit more so it's easier to place it. If the spacing is instantly expanded it's very possible to feel like you lost your place when everything is instantly offset from where it was which adds to the need of animating it out as smoothly as it can.

If there's no way to do something like this, I might have to resort to interpolating it at a fast enough speed where it doesn't visually stair-step as much, I might even consider having it overshoot so it can have a small "bounce" feel in the space expansion.

**Comment by ocornut** [comment]
Unfortunately, correctly removing the rounding during layout is going to be a huge task on my end, so it's not going to be solved from this angle soon. 

Manually setting the cursor between each item using `SetCursorScreenPos()` might be the easiest viable workaround for you, if you believe that's worth the hassle (I imagine it's not).

Stair-stepping is mostly going to be jarring if you have multiple gaps widening simultaneously, as each of them will act as a multiplier. If you only have one widening (while perhaps the previous one is shrinking) it may visually acceptable

**Comment by kraziemaniac** [comment]
Oh, that's interesting. I hadn't considered setting the cursor position as an option that might produce the results I need. If it in itself doesn't cause the stair-stepping, then it could be worth it if it can end up feeling good. I might try an easy proof of concept test to see it this might work in my case.

Also limiting the gaps to one or maybe even 2 at any given moment really might lessen things, thanks for the ideas and tips!

**Comment by kraziemaniac** [comment]
I've been able to try a few things, and setting the cursor pos did give me the visual smoothness that I was looking for. Granted I had to structure things a particular way, but I believe this is the way that's going to work in this case, and it also showed me different ways to work with Imgui if I really need to do more interesting things.

For clarity on my end solution I've decided to set the cursor pos out of place at least twice for all tree nodes then returned it somewhere proper when it's relevant for that tree node. The first time is so that I can create a child that overlaps part of the current and next tree node header so I can use that child shape as a hover check. When that hover check succeeds I have that request for a "wedge" to be placed between the relevant 2 tree nodes. This "wedge" will set the cursor pos between tree nodes so the 2nd tree node in that pair can be visually offset smoothly. Outside of this there will be a child placed in between both tree nodes that'll be used as a drag target that'll expand itself at the same time, but it's going to stairstep in it's expansion, but that shouldn't matter too much since things look correct.

It's a fair amount of things to do for this, but it's starting to feel good and I have the time currently to implement it this way.

Thanks for the tips again!
